### PR TITLE
Implement AddApplication API endpoint and UI

### PR DESCRIPTION
**Pull request by jpellizzari**
Closes: #634 
Closes: #736

**EDIT**: #829 was uncovered in this PR. I am planning on merging this in its current state, as #829 is not unique to this PR or the `AddApplication` method.

Adds a preliminary "Add" function to the API and UI. This is very much a first draft, as we are currently re-thinking how `Application` repositories work. In addition, an `Application` does not have an interstitial state after it is pull requested, making the UI very limited (same for the CLI). 

Helm is not currently supported by the UI or API as of this PR.

Here is the flow:

1. User clicks "Add Application"
Screenshot from 2021-09-28 13-44-39

2. User adds the required inputs to the form:

Screenshot from 2021-09-28 13-12-47

3. After submitting:

Screenshot from 2021-09-28 10-39-38

 Error state:
Screenshot from 2021-09-28 13-47-44

### CODE REVIEW

File: pkg/apputils/utils.go [hunk 1]
```diff
@@ -23,6 +23,7 @@ import (
 type AppFactory interface {
 	GetKubeService() (kube.Kube, error)
 	GetAppService(ctx context.Context, name, namespace string) (app.AppService, error)
+	GetAppServiceForAdd(ctx context.Context, url, configUrl, namespace string, isHelmRepository bool, dryRun bool) (app.AppService, error)
```
Comment (bigkevmcd) [inline pkg/apputils/utils.go:26]: Shouldn't this be documented with the expected behaviour of implementations of the interface?

File: pkg/server/server.go [hunk 2]
```diff
@@ -293,6 +293,54 @@ func mapSourceSpecToReponse(src client.Object) *pb.Source {
 	return source
 }
 
+func (s *applicationServer) AddApplication(ctx context.Context, msg *pb.AddApplicationRequest) (*pb.AddApplicationResponse, error) {
+	appUrl, err := gitproviders.NewNormalizedRepoURL(msg.Url)
+	if err != nil {
+		return nil, grpcStatus.Error(codes.InvalidArgument, fmt.Sprintf("unable to parse app url %q: %s", msg.Url, err))
```
Comment (bigkevmcd) [inline pkg/server/server.go:299]: I suspect you could use https://pkg.go.dev/google.golang.org/grpc/status#Errorf here?

There are other cases where you're using `fmt.Sprintf` when you could use `Errorf`

File: pkg/server/server.go [hunk 3]
```diff
@@ -293,6 +293,54 @@ func mapSourceSpecToReponse(src client.Object) *pb.Source {
...
+
+	var configUrl gitproviders.NormalizedRepoURL
+	if msg.ConfigUrl != "" {
+		configUrl, err = gitproviders.NewNormalizedRepoURL(msg.ConfigUrl)
+		if err != nil {
+			return nil, grpcStatus.Error(codes.InvalidArgument, fmt.Sprintf("unable to parse config url %q: %s", msg.ConfigUrl, err))
+		}
+	}
+
+	appSrv, err := s.appFactory.GetAppServiceForAdd(ctx, appUrl.String(), configUrl.String(), msg.Namespace, false, false)
```
Comment (bigkevmcd) [inline pkg/server/server.go:310]: There's a smell here, that you have `false, false`?

Is there a better API call that doesn't need it?

File: pkg/apputils/utils.go [hunk 4]
```diff
@@ -32,6 +33,10 @@ func (f *DefaultAppFactory) GetAppService(ctx context.Context, name, namespace s
 	return GetAppService(ctx, name, namespace)
 }
 
+func (f *DefaultAppFactory) GetAppServiceForAdd(ctx context.Context, url, configUrl, namespace string, isHelmRepository bool, dryRun bool) (app.AppService, error) {
```
Comment (bigkevmcd) [inline pkg/apputils/utils.go:36]: And maybe this could be documented?

File: ui/hooks/common.ts [hunk 5]
```diff
@@ -22,7 +22,7 @@ export function useRequestState<T>(): [
 ] {
   const [state, setState] = useState<RequestState<T>>({
     value: null,
-    loading: true,
+    loading: false,
```
Comment (jrryjcksn) [inline ui/hooks/common.ts:25]: Just for my edification, why did this change?

File: ui/pages/ApplicationDetail.tsx [hunk 6]
```diff
@@ -74,37 +79,50 @@ function ApplicationDetail({ className, name }: Props) {
       {authSuccess && (
         <Alert severity="success" message="Authentication Successful" />
       )}
+      {error && (
+        <Alert
+          severity="error"
+          title="Error fetchign Application"
```
Comment (jrryjcksn) [inline ui/pages/ApplicationDetail.tsx:85]: Typo

**Review by jrryjcksn — APPROVED** [review]
LGTM. A few minor issues (the `DeploymentType` => `SourceType`) is important, though.

File: ui/hooks/common.ts (hunk 5 above)
Comment (jpellizzari) [reply ui/hooks/common.ts:25]: Setting `loading` to true when the hook is instantiated isn't really accurate. Loading should get set when the request gets started.

File: pkg/server/server.go (hunk 3 above)
Comment (jpellizzari) [reply pkg/server/server.go:310]: > There's a smell here, that you have false, false?

Agreed. Let me see if I can improve it.

File: pkg/apputils/utils.go (hunk 1 above)
Comment (jpellizzari) [reply pkg/apputils/utils.go:26]: We haven't been enforcing that on this project. I can add a doc comment to this `interface`, but the real solution will be to turn on the linting options for doc comments, then documenting everything at once.

File: api/v1alpha1/application_types.go [hunk 7]
```diff
@@ -87,6 +87,10 @@ type Application struct {
 	Status ApplicationStatus `json:"status,omitempty"`
 }
 
+func (a *Application) IsHelmRepository() bool {
+	return a.Spec.DeploymentType == DeploymentTypeHelm
```
[the remark this answers is not in the document]
Comment (jpellizzari) [reply api/v1alpha1/application_types.go]: Fixed

File: ui/pages/ApplicationDetail.tsx (hunk 6 above)
Comment (jpellizzari) [reply ui/pages/ApplicationDetail.tsx]: Fixed! Nice catch

File: pkg/server/server_test.go [hunk 8]
```diff
@@ -540,6 +539,85 @@ var _ = Describe("ApplicationsServer", func() {
...
+			gp.GetRepoVisibilityStub = func(s string) (*gitprovider.RepositoryVisibility, error) {
+				return gitprovider.RepositoryVisibilityVar(gitprovider.RepositoryVisibilityInternal), nil
+			}
+
+			gp.CreatePullRequestToUserRepoStub = func(urr gitprovider.UserRepositoryRef, s1, s2 string, cf []gitprovider.CommitFile, s3, s4, s5 string) (gitprovider.PullRequest, error) {
+				return testutils.DummyPullRequest{}, nil
+			}
+
+			res, err := appsClient.AddApplication(contextWithAuth(ctx), appRequest)
+			Expect(err).NotTo((HaveOccurred()))
```
[the remark this answers is not in the document]
Comment (jpellizzari) [reply pkg/server/server_test.go]: Fixed

File: pkg/server/server_test.go [hunk 9]
```diff
@@ -540,6 +539,85 @@ var _ = Describe("ApplicationsServer", func() {
...
+			Expect(res.Success).To(BeTrue())
+
+			Expect(gp.CreatePullRequestToUserRepoCallCount()).To(BeNumerically(">", 0), "should have made a PR")
```
[the remark this answers is not in the document]
Comment (jpellizzari) [reply pkg/server/server_test.go]: Good call. Fixed.

File: pkg/server/server.go (hunk 3 above)
Comment (jpellizzari) [reply pkg/server/server.go]: Changed to use a `struct` as an argument to `GetAppServiceForAdd`. There was a previous effort to refactor the `apputils`, but it ended up causing more problems than it solved and was closed: https://github.com/weaveworks/weave-gitops/pull/788

File: pkg/server/server.go (hunk 2 above)
Comment (jpellizzari) [reply pkg/server/server.go]: Nice catch. Fixed to use `Errorf` in that file.

File: pkg/server/server_test.go [hunk 10]
```diff
@@ -540,6 +539,85 @@ var _ = Describe("ApplicationsServer", func() {
...
+
+			gp.CreatePullRequestToUserRepoStub = func(urr gitprovider.UserRepositoryRef, s1, s2 string, cf []gitprovider.CommitFile, s3, s4, s5 string) (gitprovider.PullRequest, error) {
+				return testutils.DummyPullRequest{}, nil
+			}
+
+			res, err := appsClient.AddApplication(contextWithAuth(ctx), appRequest)
+			Expect(err).NotTo((HaveOccurred()))
+			Expect(res.Success).To(BeTrue())
+
+			Expect(configGit.CommitCallCount()).To(BeNumerically(">", 0), "should have committed to config git repo")
```
[the remark this answers is not in the document]
Comment (jpellizzari) [reply pkg/server/server_test.go]: Fixed

File: pkg/server/server_test.go [hunk 11]
```diff
@@ -540,6 +539,85 @@ var _ = Describe("ApplicationsServer", func() {
...
+				Namespace: namespace.Name,
+				Url:       "ssh://git@github.com/some-org/somerepo.git",
+				Path:      "./k8s/mydir",
+				Branch:    "main",
+			}
+			gp.GetRepoVisibilityStub = func(s string) (*gitprovider.RepositoryVisibility, error) {
+				return gitprovider.RepositoryVisibilityVar(gitprovider.RepositoryVisibilityInternal), nil
+			}
+
+			gp.CreatePullRequestToUserRepoStub = func(urr gitprovider.UserRepositoryRef, s1, s2 string, cf []gitprovider.CommitFile, s3, s4, s5 string) (gitprovider.PullRequest, error) {
```
Comment (luizbafilho) [inline pkg/server/server_test.go:557]: For just setting up the return you could use this syntax:

`gp.CreatePullRequestToUserRepoReturns(testutils.DummyPullRequest{}, nil)`

It should help cleaning up a bit all those mocks

File: pkg/server/server_test.go [hunk 12]
```diff
@@ -653,9 +729,9 @@ var _ = Describe("ApplicationsServer", func() {
 				}
 
 				appFactory := &apputilsfakes.FakeAppFactory{}
-				appFactory.GetAppServiceStub = func(ctx context.Context, name, namespace string) (app.AppService, error) {
-					return app.New(ctx, nil, nil, nil, nil, nil, kubeClient, nil), nil
-				}
+				// appFactory.GetAppServiceStub = func(ctx context.Context, name, namespace string) (app.AppService, error) {
```
Comment (luizbafilho) [inline pkg/server/server_test.go:732]: clean this up

File: ui/pages/ApplicationDetail.tsx [hunk 13]
```diff
@@ -74,37 +79,50 @@ function ApplicationDetail({ className, name }: Props) {
...
-        app={app}
+        app={application}
         onAuthClick={() => setGithubAuthModalOpen(true)}
       />
       <GithubDeviceAuthModal
         onSuccess={() => {
           setAuthSuccess(true);
           // Get CommitsTable to retry after auth
-          setApp({ ...app });
+          // setApp({ ...application });
```
Comment (luizbafilho) [inline ui/pages/ApplicationDetail.tsx:123]: clean this up

File: pkg/server/server_test.go (hunk 11 above)
Comment (luizbafilho) [inline pkg/server/server_test.go:557]: For just setting up the return you could use this syntax:

`gp.CreatePullRequestToUserRepoReturns(testutils.DummyPullRequest{}, nil)`

It should help cleaning up a bit all those mocks

**Review by luizbafilho — COMMENTED** [review]
looks good just minor to clean up commented code and a way to simplify the mocking setup which get quit ugly when the method has too many params.

File: pkg/server/server_test.go (hunk 11 above)
Comment (jpellizzari) [reply pkg/server/server_test.go:557]: Did not realize that was an option. Not sure why we ever used the full signature.

File: pkg/server/server_test.go (hunk 12 above)
Comment (jpellizzari) [reply pkg/server/server_test.go:732]: > clean this up

Is there a nicer way to say this?

Comment (luizbafilho) [reply pkg/server/server_test.go:732]: sorry, didn't mean to be rude I wrote it more like an #TODO comment.

File: pkg/server/server_test.go (hunk 11 above)
Comment (jpellizzari) [reply pkg/server/server_test.go]: Updated to the cleaner funcs

File: pkg/apputils/utils.go (hunk 1 above)
Comment (jpellizzari) [reply pkg/apputils/utils.go]: Documented

**Review by luizbafilho — APPROVED** [review]

**Pull request merged**